lib/common/xml.c

```diff
@@ -1498,6 +1523,70 @@ xml_apply_patchset_v2(xmlNode *xml, xmlNode *patchset)
             continue;
 
         } else if(strcmp(op, "create") == 0) {
+            // Delay the adding of the "create" object
+            changes = g_list_append(changes, change);
+
+        } else if(strcmp(op, "move") == 0) {
+            changes = g_list_append(changes, change);
+
+            // Temporarily put the "move" object after the last sibling
+            if (match->parent != NULL && match->parent->last != NULL) {
+                xmlAddNextSibling(match->parent->last, match);
+            }
+
+        } else if(strcmp(op, "delete") == 0) {
+            free_xml(match);
+
+        } else if(strcmp(op, "modify") == 0) {
+            xmlAttr *pIter = pcmk__first_xml_attr(match);
+            xmlNode *attrs = __xml_first_child(first_named_child(change, XML_DIFF_RESULT));
+
+            if(attrs == NULL) {
+                rc = -ENOMSG;
+                continue;
+            }
+            while(pIter != NULL) {
+                const char *name = (const char *)pIter->name;
+
+                pIter = pIter->next;
+                xml_remove_prop(match, name);
+            }
+
+            for (pIter = pcmk__first_xml_attr(attrs); pIter != NULL; pIter = pIter->next) {
+                const char *name = (const char *)pIter->name;
+                const char *value = crm_element_value(attrs, name);
+
+                crm_xml_add(match, name, value);
+            }
+
+        } else {
+            crm_err("Unknown operation: %s", op);
+        }
+    }
+
+    // Changes should be generated in the right order. Double checking.
+    changes = g_list_sort(changes, sort_change_by_position);
+
+    for (gIter = changes; gIter; gIter = gIter->next) {
+        xmlNode *match = NULL;
+        const char *op = NULL;
+        const char *xpath = NULL;
+        int position = -1;
+
+        change = gIter->data;
+
+        op = crm_element_value(change, XML_DIFF_OP);
+        xpath = crm_element_value(change, XML_DIFF_PATH);
+
+        crm_trace("Continue processing %s %s", change->name, op);
+        if (op == NULL) {
+            continue;
+        }
+
+        match = __xml_find_path(xml, xpath, position);
```

Comment:
Given how often patchsets are passed around, I'm wondering about the impact of making this call twice. It would be a bit of a pain, but I'm thinking the changes list should have struct { xmlNode *change, xmlNode *match } items. You wouldn't need to grab xpath in that case either.

Reply:
Good idea.